File: my_openclaw/mini_agent_ts/src/http-client.ts
/**
 * http-client.ts — HTTP client for OpenAI-compatible streaming API
 *
 * Borrowed from openclaw's API call patterns:
 *   - src/plugins/provider-zai-endpoint.ts  (URL construction: `${baseUrl}/chat/completions`)
 *   - src/agents/tools/image-tool.test.ts   (fetch + SSE pattern)
 *   - src/utils/fetch-timeout.ts            (fetchWithTimeout)
 *
 * In openclaw, the actual LLM API call happens deep inside the
 * @mariozechner/pi-ai library, but the URL construction and auth
 * header patterns are visible in the provider endpoint probing code.
 *
 * The SSE (Server-Sent Events) streaming protocol:
 *   - Response is a stream of lines
 *   - Lines starting with "data: " contain JSON payloads
 *   - "data: [DONE]" signals stream end
 *   - Each payload has: choices[0].delta.content (the token)
 *
 * Data flow:
 *   fetch(POST) → ReadableStream → TextDecoder → line split → SSE parse → token callback
 */

import type { ChatMessage } from "./message.js";

// ── Types ─────────────────────────────────────────────────────────

export type TokenCallback = (token: string, done: boolean) => void;

export interface StreamChatOptions {
  baseUrl: string;
  apiKey: string;
  model: string;
  messages: ChatMessage[];
  onToken: TokenCallback;
  timeoutMs?: number;
}

// ── URL construction ──────────────────────────────────────────────
// Mirrors openclaw's pattern: `${params.baseUrl}/chat/completions`
// with trailing-slash stripping and http→https upgrade.

function buildUrl(baseUrl: string): string {
  let url = baseUrl.replace(/\/+$/, "");

  // Auto-upgrade http:// to https:// — most LLM API providers require HTTPS.
  // Plain HTTP often gets a 404 from the gateway even though the route exists.
  if (url.startsWith("http://")) {
    url = "https://" + url.slice(7);
  }

  return `${url}/chat/completions`;
}

// ── SSE line parser ───────────────────────────────────────────────
// Parses Server-Sent Events format line by line.
// Each "data: {...}" line contains a JSON chunk with:
//   { choices: [{ delta: { content: "token" } }] }

function processSSELine(line: string, onToken: TokenCallback): void {
  // Skip empty lines and comments
  if (!line || line.startsWith(":")) return;

  // We only care about "data: " prefixed lines
  if (!line.startsWith("data: ")) return;

  const payload = line.slice(6);

  // Stream termination signal
  if (payload === "[DONE]") {
    onToken("", true);
    return;
  }

  // Parse JSON chunk — navigate: choices[0].delta.content
  try {
    const chunk = JSON.parse(payload) as {
      choices?: Array<{
        delta?: { content?: string; reasoning_content?: string };
        finish_reason?: string | null;
      }>;
    };

    const delta = chunk.choices?.[0]?.delta;
    if (delta?.content) {
      onToken(delta.content, false);
    }
    // Some models (DeepSeek) emit reasoning_content for chain-of-thought
    if (delta?.reasoning_content) {
      onToken(delta.reasoning_content, false);
    }
  } catch {
    // Malformed JSON — skip silently (mirrors openclaw's defensive parsing)
  }
}

// ── streamChat ────────────────────────────────────────────────────
// Main API call function. Uses Node.js native fetch (available since v18).
//
// Request body mirrors openclaw's format:
//   { "model": "...", "stream": true, "messages": [...] }
//
// Auth header: "Authorization: Bearer {api_key}"

export async function streamChat(opts: StreamChatOptions): Promise<void> {
  const url = buildUrl(opts.baseUrl);
  const timeoutMs = opts.timeoutMs ?? 120_000;

  // Build request body — same format as openclaw's OpenAI-compatible calls
  const body = JSON.stringify({
    model: opts.model,
    stream: true,
    messages: opts.messages,
  });

  process.stderr.write(`[http] POST ${url}  model=${opts.model}\n`);

  // Create abort controller for timeout
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${opts.apiKey}`,
      },
      body,
      signal: controller.signal,
    });

    if (!response.ok) {
      // Try to read error body for diagnostics
      let errorBody = "";
      try {
        errorBody = await response.text();
      } catch { /* ignore */ }
      throw new Error(
        `HTTP ${response.status} from ${url}${errorBody ? `: ${errorBody}` : ""}`,
      );
    }

    if (!response.body) {
      throw new Error("Response body is null — streaming not supported?");
    }

    // ── Stream processing ───────────────────────────────────────
    // Read the response as a stream of bytes, decode to text,
    // split into lines, and process each SSE line.
    //
    // This mirrors the C version's write_callback approach but uses
    // Node.js ReadableStream + TextDecoder.

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      // Split on newlines and process complete lines
      const lines = buffer.split("\n");
      // Keep the last (possibly incomplete) line in the buffer
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        const trimmed = line.replace(/\r$/, "");
        processSSELine(trimmed, opts.onToken);
      }
    }

    // Process any remaining data in buffer
    if (buffer.trim()) {
      processSSELine(buffer.replace(/\r$/, ""), opts.onToken);
    }
  } finally {
    clearTimeout(timer);
  }
}

File: my_openclaw/mini_agent/http_client.c
/*
 * http_client.c - HTTP client using libcurl for SSE streaming
 *
 * This is the network layer of our mini agent. It mirrors openclaw's
 * approach to calling OpenAI-compatible endpoints:
 *
 *   POST {base_url}/chat/completions
 *   Authorization: Bearer {api_key}
 *   Content-Type: application/json
 *   Body: { "model": "...", "stream": true, "messages": [...] }
 *
 * The response is Server-Sent Events (SSE):
 *   data: {"choices":[{"delta":{"content":"Hello"}}]}
 *   data: [DONE]
 *
 * We parse each SSE line in the write callback and extract the
 * content delta, forwarding it to the user's token callback.
 */
#include "http_client.h"
#include "json_parser.h"
#include <curl/curl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ── SSE stream parsing context ────────────────────────────────── */

typedef struct {
    stream_token_cb callback;
    void           *userdata;
    char           *line_buf;    /* accumulates partial lines */
    size_t          line_len;
    size_t          line_cap;
} sse_ctx_t;

static void sse_ctx_init(sse_ctx_t *ctx, stream_token_cb cb, void *ud)
{
    ctx->callback = cb;
    ctx->userdata = ud;
    ctx->line_cap = 1024;
    ctx->line_buf = malloc(ctx->line_cap);
    ctx->line_len = 0;
}

static void sse_ctx_free(sse_ctx_t *ctx)
{
    free(ctx->line_buf);
}

/*
 * Process one complete SSE line.
 * Lines look like:
 *   data: {"id":"...","choices":[{"delta":{"content":"token"}}]}
 *   data: [DONE]
 */
static void process_sse_line(sse_ctx_t *ctx, const char *line)
{
    /* Skip empty lines and comments */
    if (line[0] == '\0' || line[0] == ':')
        return;

    /* We only care about "data: " prefixed lines */
    if (strncmp(line, "data: ", 6) != 0)
        return;

    const char *payload = line + 6;

    /* Check for stream termination */
    if (strcmp(payload, "[DONE]") == 0) {
        ctx->callback("", 0, 1, ctx->userdata);
        return;
    }

    /* Parse the JSON chunk */
    json_value_t *root = json_parse(payload);
    if (!root) return;

    /*
     * Navigate: choices[0].delta.content
     * This is the standard OpenAI streaming response format.
     */
    json_value_t *choices = json_object_get(root, "choices");
    if (choices && choices->type == JSON_ARRAY && choices->u.array.count > 0) {
        json_value_t *first = choices->u.array.items[0];
        json_value_t *delta = json_object_get(first, "delta");
        if (delta) {
            const char *content = json_get_string(delta, "content");
            if (content && content[0] != '\0') {
                ctx->callback(content, strlen(content), 0, ctx->userdata);
            }
        }
    }

    json_free(root);
}

/* ── libcurl write callback ────────────────────────────────────── */

static size_t write_callback(char *ptr, size_t size, size_t nmemb,
                             void *userdata)
{
    size_t total = size * nmemb;
    sse_ctx_t *ctx = (sse_ctx_t *)userdata;

    for (size_t i = 0; i < total; i++) {
        char c = ptr[i];
        if (c == '\n') {
            /* End of line - process it */
            ctx->line_buf[ctx->line_len] = '\0';
            /* Strip trailing \r */
            if (ctx->line_len > 0 && ctx->line_buf[ctx->line_len - 1] == '\r')
                ctx->line_buf[--ctx->line_len] = '\0';
            process_sse_line(ctx, ctx->line_buf);
            ctx->line_len = 0;
        } else {
            /* Accumulate character */
            if (ctx->line_len + 1 >= ctx->line_cap) {
                ctx->line_cap *= 2;
                ctx->line_buf = realloc(ctx->line_buf, ctx->line_cap);
            }
            ctx->line_buf[ctx->line_len++] = c;
        }
    }

    return total;
}

/* ── Public API ────────────────────────────────────────────────── */

void http_init(void)
{
    curl_global_init(CURL_GLOBAL_DEFAULT);
}

void http_cleanup(void)
{
    curl_global_cleanup();
}

int http_stream_chat(const char *base_url,
                     const char *api_key,
                     const char *model,
                     const char *messages_json,
                     stream_token_cb callback,
                     void *userdata)
{
    CURL *curl = curl_easy_init();
    if (!curl) {
        fprintf(stderr, "[http] curl_easy_init failed\n");
        return -1;
    }

    /* ── Build URL: {base_url}/chat/completions ────────────────── */
    char url[512];
    /* Strip trailing slash from base_url */
    size_t blen = strlen(base_url);
    while (blen > 0 && base_url[blen - 1] == '/')
        blen--;

    /*
     * Auto-upgrade http:// to https:// for security.
     * Most LLM API providers require HTTPS; plain HTTP often gets
     * a 404 from the gateway even though the route exists.
     */
    char upgraded_base[512];
    if (blen >= 7 && strncmp(base_url, "http://", 7) == 0) {
        snprintf(upgraded_base, sizeof(upgraded_base), "https://%.*s",
                 (int)(blen - 7), base_url + 7);
        blen = strlen(upgraded_base);
        snprintf(url, sizeof(url), "%s/chat/completions", upgraded_base);
    } else {
        snprintf(url, sizeof(url), "%.*s/chat/completions",
                 (int)blen, base_url);
    }

    fprintf(stderr, "[http] POST %s  model=%s\n", url, model);

    /* ── Build request body ────────────────────────────────────── */
    /*
     * {
     *   "model": "<model>",
     *   "stream": true,
     *   "messages": [ ... ]
     * }
     */
    size_t body_cap = strlen(messages_json) + strlen(model) + 128;
    char *body = malloc(body_cap);
    snprintf(body, body_cap,
             "{\"model\":\"%s\",\"stream\":true,\"messages\":%s}",
             model, messages_json);

    /* ── Set headers ───────────────────────────────────────────── */
    struct curl_slist *headers = NULL;
    headers = curl_slist_append(headers, "Content-Type: application/json");

    char auth_header[320];
    snprintf(auth_header, sizeof(auth_header),
             "Authorization: Bearer %s", api_key);
    headers = curl_slist_append(headers, auth_header);

    /* ── Configure curl ────────────────────────────────────────── */
    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_POST, 1L);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body);
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 120L);
    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 10L);

    /* SSE streaming: use our write callback */
    sse_ctx_t ctx;
    sse_ctx_init(&ctx, callback, userdata);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &ctx);

    /* ── Execute ───────────────────────────────────────────────── */
    CURLcode res = curl_easy_perform(curl);

    int ret = 0;
    if (res != CURLE_OK) {
        fprintf(stderr, "[http] curl error: %s\n",
                curl_easy_strerror(res));
        ret = -1;
    } else {
        /* Check HTTP status */
        long http_code = 0;
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
        if (http_code != 200) {
            fprintf(stderr, "[http] HTTP %ld from %s\n", http_code, url);
            ret = -1;
        }
    }

    /* ── Cleanup ───────────────────────────────────────────────── */
    sse_ctx_free(&ctx);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    free(body);

    return ret;
}
